include/swift/Parse/ParsedRawSyntaxRecorder.h

```diff
@@ -0,0 +1,72 @@
+//===--- ParsedRawSyntaxRecorder.h - Raw Syntax Parsing Recorder ----------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2019 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+//
+// This file defines the ParsedRawSyntaxRecorder, which is the interface the
+// parser is using to pass parsed syntactic elements to a SyntaxParseActions
+// receiver and get a ParsedRawSyntaxNode object back.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef SWIFT_PARSE_PARSEDRAWSYNTAXRECORDER_H
+#define SWIFT_PARSE_PARSEDRAWSYNTAXRECORDER_H
+
+#include "swift/Basic/LLVM.h"
+#include <memory>
+
+namespace swift {
+
+class ParsedRawSyntaxNode;
+class SyntaxParseActions;
+class SourceLoc;
+class Token;
+enum class tok;
+
+namespace syntax {
+  enum class SyntaxKind;
+  struct Trivia;
+}
+
+class ParsedRawSyntaxRecorder {
+  std::shared_ptr<SyntaxParseActions> SPActions;
+
+public:
+  explicit ParsedRawSyntaxRecorder(std::shared_ptr<SyntaxParseActions> spActions)
+    : SPActions(std::move(spActions)) {}
+
+  ParsedRawSyntaxNode recordToken(const Token &tok,
+                                  const syntax::Trivia &leadingTrivia,
+                                  const syntax::Trivia &trailingTrivia);
+
+  /// Record a missing token. \c loc can be invalid or an approximate location
+  /// of where the token would be if not missing.
+  ParsedRawSyntaxNode recordMissingToken(tok tokenKind, SourceLoc loc);
+
+  /// The provided \c elements are an exact layout appropriate for the syntax
+  /// \c kind. Missing optional elements are represented with a null
+  /// ParsedRawSyntaxNode object.
+  ParsedRawSyntaxNode recordExactRawSyntax(syntax::SyntaxKind kind,
+                                     ArrayRef<ParsedRawSyntaxNode> elements);
```

Comment:
`recordRawSyntax` is good enough IMO. We don't have a non-exact version of recording and it doesn't make much sense.

Reply:
Done.